data/src/main/java/com/microsoft/azure/kusto/data/Utils.java

```diff
@@ -81,12 +76,156 @@ static String post(String url, String payload, InputStream stream, Integer timeo
+        return postToStreamingOutput(url, payload, timeoutMs, headers, 0);
+    }
+
+    static InputStream postToStreamingOutput(String url, String payload, long timeoutMs, Map<String, String> headers, int redirectCount) throws DataServiceException, DataClientException {
+        long timeoutTimeMs = System.currentTimeMillis() + timeoutMs;
+        URI uri = parseUriFromUrlString(url);
+        boolean returnInputStream = false;
+        /*
+         *  The caller must close the inputStream to close the following underlying resources (httpClient and httpResponse).
+         *  We use CloseParentResourcesStream so that when the stream is closed, these resources are closed as well. We
+         *  shouldn't need to do that, per https://hc.apache.org/httpcomponents-client-4.5.x/current/tutorial/html/fundamentals.html:
+         *  "In order to ensure proper release of system resources one must close either the content stream associated
+         *  with the entity or the response itself." However, in my testing this wasn't reliably the case.
+         *  We further use EofSensorInputStream to close the stream even if not explicitly closed, once all content is consumed.
+         */
+        CloseableHttpClient httpClient = null;
+        CloseableHttpResponse httpResponse = null;
+        try {
+            httpClient = getHttpClient(Math.toIntExact(timeoutTimeMs - System.currentTimeMillis()));
+            httpResponse = httpClient.execute(setupHttpPostRequest(uri, payload, null, headers));
+
+            int responseStatusCode = httpResponse.getStatusLine().getStatusCode();
+
+            if (responseStatusCode == HttpStatus.SC_OK) {
+                InputStream contentStream = new EofSensorInputStream(new CloseParentResourcesStream(httpClient, httpResponse), null);
+                Optional<Header> contentEncoding = Arrays.stream(httpResponse.getHeaders(HttpHeaders.CONTENT_ENCODING)).findFirst();
+                if (contentEncoding.isPresent()) {
+                    if (contentEncoding.get().getValue().contains("gzip")) {
+                        GZIPInputStream gzipInputStream = new GZIPInputStream(contentStream);
+                        returnInputStream = true;
+                        return gzipInputStream;
+                    } else if (contentEncoding.get().getValue().contains("deflate")) {
+                        DeflaterInputStream deflaterInputStream = new DeflaterInputStream(contentStream);
+                        returnInputStream = true;
+                        return deflaterInputStream;
+                    }
+                }
+                // Though the server responds with a gzip/deflate Content-Encoding header, we reach here because httpclient uses LazyDecompressingStream which handles the above logic
+                returnInputStream = true;
+                return contentStream;
+            }
+
```

Comment:
As talked - we should expose the internal exception 
Add here 
            DataWebException exc = new DataWebException(EntityUtils.toString(httpResponse.getEntity()), httpResponse);
and throw it later